src/common/host_device_vector.cu

```diff
@@ -35,13 +36,21 @@ struct HostDeviceVectorImpl {
   void operator=(const HostDeviceVectorImpl<T>&) = delete;
   void operator=(HostDeviceVectorImpl<T>&&) = delete;
 
-  size_t Size() const { return on_d_ ? data_d_.size() : data_h_.size(); }
+  size_t Size() {
+     if (on_d_) {
+       if (data_d_ == nullptr) data_d_.reset(new thrust::device_vector<T>());
```

Comment:
Maybe just assume that `data_d_` is initialized if `on_d_` is true. In this way, many unnecessary checks can be avoided.